### PR TITLE
#17 - Set Up Lambda

### PR DESCRIPTION
**Pull request by iamstevedavis**
Setup a lambda to authenticate and hit @clark-teeple new route. 
Setup a github workflow to push the built lambda to S3.
 - We can make all infrastructure (s3, lambda) in cloud formation, not sure if it supports this auto push though.

Used native HTTP to avoid having to npm install and push the node_modules bundle to the lambda.

### CODE REVIEW

File: .github/workflows/workflow.yml [hunk 1]
```diff
@@ -0,0 +1,27 @@
+name: Upload App
+
+on:
+  push:
+    branches:
+    - master
```
Comment (clark-teeple) [inline .github/workflows/workflow.yml:6]: should we do a release branch?

File: lambda/lambda.js [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    }
+}
+
+  return new Promise((resolve, reject) => {
+    const lib = baseUrl.startsWith("https") ? require("https") : require("http");
+    const request = lib.get(options, (res) => {
+      if (res.statusCode < 200 || res.statusCode > 299) {
+        return reject(`SendMessage API Call Failed with Code: ${res.statusCode}`);
+      }
+      // Not interested in the response.
```
Comment (clark-teeple) [inline lambda/lambda.js:57]: could we turf the response off to cloudwatch or something? we don't currently have any visibility into failed messages.

Comment (clark-teeple) [reply lambda/lambda.js:57]: could handle in the API just as easily though if we started a log account somewhere or something.

Comment (iamstevedavis) [reply lambda/lambda.js:57]: We could but that means the lambda might sit waiting and running for the API to process all those messages and send a response

Comment (clark-teeple) [reply lambda/lambda.js:57]: Oh true. Alright, we'll do it somewhere else.

**Comment by garethbdavis** [comment]
It looks like you're uploading the lambda code to an S3 bucket.  This alone won't update the function.  You'd need to use something like https://docs.aws.amazon.com/cli/latest/reference/lambda/update-function-code.html.

We're using Terraform rather than cloud formation to create infrastructure, but updating lambda code would fall outside of what we'd control with terraform.

For the MVP, it may be worth just using a Makefile to deploy the Lambda.  That way we don't have to worry about branching strategy and AWS access.

**Comment by iamstevedavis** [comment]
@garethbdavis I know it won't update the function, we can copy the s3 filepath to the lambda for now.  Regarding your slack comment, it would be great if you could update terraform for the lambda.

**Comment by garethbdavis** [comment]
@iamstevedavis Sounds good.  I'll take a look at terraform for the Lambda.  Could you give me any requirements it has, either here on in an issue:

- trigger type (cron?)
- any non-default settings for memory, timeout etc
- anything else I should know

**Comment by Wiles** [comment]
Closed since we decided to go with the cron approach instead of using Lambda.

**Pull request closed without merging**